File: 0054-spiral-matrix/0054-spiral-matrix.js
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
const spiralOrder = matrix => {
    let top = left = 0;
    let bottom = matrix.length - 1;
    let right = matrix[0].length - 1;
    let direction = 'right';
    const result = [];
    while (left <= right && top <= bottom) {
        if (direction === 'right') {
            for (let i = left; i <= right; i++) {
                result.push(matrix[top][i]);
            }
            direction = 'down';
            top += 1;
        } else if (direction === 'down') {
            for (let i = top; i <= bottom; i++) {
                result.push(matrix[i][right]);
            }
            direction = 'left';
            right -= 1;
        } else if (direction === 'left') {
            for (let i = right; i >= left; i--) {
                result.push(matrix[bottom][i]);
            }
            direction = 'up';
            bottom -= 1;
        } else {
            for (let i = bottom; i >= top; i--) {
                result.push(matrix[i][left]);
            }
            direction = 'right';
            left += 1;
        }
    }
    return result;
};

File: 0054-spiral-matrix/0054-spiral-matrix.rb
# @param {Integer[][]} matrix
# @return {Integer[]}
def spiral_order(matrix)
    left = 0
    right = matrix[0].length
    top = 0
    bottom = matrix.length
    direction = 'right'
    spiral = []
    while left < right and top < bottom
        if direction == 'right'
            (left..right-1).each do |i|
                spiral.push(matrix[top][i])
            end
            top += 1
            direction = 'down'
        elsif direction == 'down'
            (top..bottom-1).each do |i|
                spiral.push(matrix[i][right-1])
            end
            right -= 1
            direction = 'left'
        elsif direction == 'left'
            (right-1).downto(left).each do |i|
                spiral.push(matrix[bottom-1][i])
            end
            bottom -= 1
            direction = 'up'
        else 
            (bottom-1).downto(top).each do |i|
                spiral.push(matrix[i][left])
            end
            left += 1
            direction = 'right'
        end
    end
    return spiral
end